File: src/main/java/com/agrifund/entities/Utilisateur.java
package com.agrifund.entities;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;

public class Utilisateur {

    private int id;
    private String nom;
    private String prenom;
    private String email;
    private String password;
    private String tel;
    private LocalDateTime dateInscrit;
    private String photo;
    private LocalDateTime derniereConnexion;
    private boolean estEnLigne;

    // Constructeurs
    public Utilisateur() {
        this.dateInscrit = LocalDateTime.now();
    }

    public Utilisateur(String nom, String prenom, String email, String password) {
        this.nom = nom;
        this.prenom = prenom;
        this.email = email;
        this.password = password;
        this.dateInscrit = LocalDateTime.now();
    }

    // Getters et Setters existants...
    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public String getNom() { return nom; }
    public void setNom(String nom) { this.nom = nom; }

    public String getPrenom() { return prenom; }
    public void setPrenom(String prenom) { this.prenom = prenom; }

    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }

    public String getPassword() { return password; }
    public void setPassword(String password) { this.password = password; }

    public String getTel() { return tel; }
    public void setTel(String tel) { this.tel = tel; }

    public LocalDateTime getDateInscrit() { return dateInscrit; }
    public void setDateInscrit(LocalDateTime dateInscrit) { this.dateInscrit = dateInscrit; }

    public String getPhoto() { return photo; }
    public void setPhoto(String photo) { this.photo = photo; }

    // Nouveaux getters/setters pour le statut en ligne
    public LocalDateTime getDerniereConnexion() { return derniereConnexion; }
    public void setDerniereConnexion(LocalDateTime derniereConnexion) {
        this.derniereConnexion = derniereConnexion;
    }

    public boolean isEstEnLigne() { return estEnLigne; }
    public void setEstEnLigne(boolean estEnLigne) { this.estEnLigne = estEnLigne; }

    /**
     * Obtenir le statut formaté (En ligne, Vu il y a X minutes, etc.)
     */
    public String getStatutEnLigne() {
        if (estEnLigne) {
            return "🟢 En ligne";
        }

        if (derniereConnexion == null) {
            return "⚪ Hors ligne";
        }

        LocalDateTime now = LocalDateTime.now();
        long minutes = ChronoUnit.MINUTES.between(derniereConnexion, now);
        long heures = ChronoUnit.HOURS.between(derniereConnexion, now);
        long jours = ChronoUnit.DAYS.between(derniereConnexion, now);

        if (minutes < 1) {
            return "🟢 En ligne";
        } else if (minutes < 60) {
            return "🕐 Vu il y a " + minutes + " min";
        } else if (heures < 24) {
            return "🕐 Vu il y a " + heures + "h";
        } else if (jours < 7) {
            return "🕐 Vu il y a " + jours + " jour" + (jours > 1 ? "s" : "");
        } else {
            return "⚪ Vu le " + derniereConnexion.format(
                    java.time.format.DateTimeFormatter.ofPattern("dd/MM/yyyy"));
        }
    }

    /**
     * Vérifier si l'utilisateur est considéré comme en ligne (actif < 5 min)
     */
    public boolean estRecentementActif() {
        if (derniereConnexion == null) return false;
        long minutes = ChronoUnit.MINUTES.between(derniereConnexion, LocalDateTime.now());
        return minutes < 5;
    }

    @Override
    public String toString() {
        return "Utilisateur{" +
                "id=" + id +
                ", nom='" + nom + '\'' +
                ", prenom='" + prenom + '\'' +
                ", email='" + email + '\'' +
                '}';
    }
}

File: src/Entity/Utilisateur.php
<?php

namespace App\Entity;

use App\Repository\UtilisateurRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: UtilisateurRepository::class)]
class Utilisateur
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 100)]
    private ?string $nom = null;

    #[ORM\Column(length: 100)]
    private ?string $prenom = null;

    #[ORM\Column(length: 255)]
    private ?string $email = null;

    #[ORM\Column(length: 255)]
    private ?string $password = null;

    #[ORM\Column(length: 20, nullable: true)]
    private ?string $tel = null;

    #[ORM\Column(nullable: true)]
    private ?\DateTime $date_inscrit = null;

    #[ORM\Column(length: 255, nullable: true)]
    private ?string $photo = null;

    #[ORM\Column]
    private ?\DateTime $derniere_connexion = null;

    #[ORM\Column(nullable: true)]
    private ?bool $est_en_ligne = null;

    /**
     * @var Collection<int, Admin>
     */
    #[ORM\OneToMany(targetEntity: Admin::class, mappedBy: 'utilsateur_id', orphanRemoval: true)]
    private Collection $idadmin;

    /**
     * @var Collection<int, Agriculteur>
     */
    #[ORM\OneToMany(targetEntity: Agriculteur::class, mappedBy: 'utilisateur_id', orphanRemoval: true)]
    private Collection $agriid;

    /**
     * @var Collection<int, Banque>
     */
    #[ORM\OneToMany(targetEntity: Banque::class, mappedBy: 'utilisateur_id', orphanRemoval: true)]
    private Collection $banqueid;

    /**
     * @var Collection<int, StatutEnLigne>
     */
    #[ORM\OneToMany(targetEntity: StatutEnLigne::class, mappedBy: 'utilisateur_id')]
    private Collection $statut;

    /**
     * @var Collection<int, HistoriqueConnexion>
     */
    #[ORM\OneToMany(targetEntity: HistoriqueConnexion::class, mappedBy: 'utilisateur_id', orphanRemoval: true)]
    private Collection $connid;

    public function __construct()
    {
        $this->idadmin = new ArrayCollection();
        $this->agriid = new ArrayCollection();
        $this->banqueid = new ArrayCollection();
        $this->statut = new ArrayCollection();
        $this->connid = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function setId(int $id): static
    {
        $this->id = $id;

        return $this;
    }

    public function getNom(): ?string
    {
        return $this->nom;
    }

    public function setNom(string $nom): static
    {
        $this->nom = $nom;

        return $this;
    }

    public function getPrenom(): ?string
    {
        return $this->prenom;
    }

    public function setPrenom(string $prenom): static
    {
        $this->prenom = $prenom;

        return $this;
    }

    public function getEmail(): ?string
    {
        return $this->email;
    }

    public function setEmail(string $email): static
    {
        $this->email = $email;

        return $this;
    }

    public function getPassword(): ?string
    {
        return $this->password;
    }

    public function setPassword(string $password): static
    {
        $this->password = $password;

        return $this;
    }

    public function getTel(): ?string
    {
        return $this->tel;
    }

    public function setTel(?string $tel): static
    {
        $this->tel = $tel;

        return $this;
    }

    public function getDateInscrit(): ?\DateTime
    {
        return $this->date_inscrit;
    }

    public function setDateInscrit(?\DateTime $date_inscrit): static
    {
        $this->date_inscrit = $date_inscrit;

        return $this;
    }

    public function getPhoto(): ?string
    {
        return $this->photo;
    }

    public function setPhoto(?string $photo): static
    {
        $this->photo = $photo;

        return $this;
    }

    public function getDerniereConnexion(): ?\DateTime
    {
        return $this->derniere_connexion;
    }

    public function setDerniereConnexion(\DateTime $derniere_connexion): static
    {
        $this->derniere_connexion = $derniere_connexion;

        return $this;
    }

    public function isEstEnLigne(): ?bool
    {
        return $this->est_en_ligne;
    }

    public function setEstEnLigne(?bool $est_en_ligne): static
    {
        $this->est_en_ligne = $est_en_ligne;

        return $this;
    }

    /**
     * @return Collection<int, Admin>
     */
    public function getIdadmin(): Collection
    {
        return $this->idadmin;
    }

    public function addIdadmin(Admin $idadmin): static
    {
        if (!$this->idadmin->contains($idadmin)) {
            $this->idadmin->add($idadmin);
            $idadmin->setUtilsateurId($this);
        }

        return $this;
    }

    public function removeIdadmin(Admin $idadmin): static
    {
        if ($this->idadmin->removeElement($idadmin)) {
            // set the owning side to null (unless already changed)
            if ($idadmin->getUtilsateurId() === $this) {
                $idadmin->setUtilsateurId(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection<int, Agriculteur>
     */
    public function getAgriid(): Collection
    {
        return $this->agriid;
    }

    public function addAgriid(Agriculteur $agriid): static
    {
        if (!$this->agriid->contains($agriid)) {
            $this->agriid->add($agriid);
            $agriid->setUtilisateurId($this);
        }

        return $this;
    }

    public function removeAgriid(Agriculteur $agriid): static
    {
        if ($this->agriid->removeElement($agriid)) {
            // set the owning side to null (unless already changed)
            if ($agriid->getUtilisateurId() === $this) {
                $agriid->setUtilisateurId(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection<int, Banque>
     */
    public function getBanqueid(): Collection
    {
        return $this->banqueid;
    }

    public function addBanqueid(Banque $banqueid): static
    {
        if (!$this->banqueid->contains($banqueid)) {
            $this->banqueid->add($banqueid);
            $banqueid->setUtilisateurId($this);
        }

        return $this;
    }

    public function removeBanqueid(Banque $banqueid): static
    {
        if ($this->banqueid->removeElement($banqueid)) {
            // set the owning side to null (unless already changed)
            if ($banqueid->getUtilisateurId() === $this) {
                $banqueid->setUtilisateurId(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection<int, StatutEnLigne>
     */
    public function getStatut(): Collection
    {
        return $this->statut;
    }

    public function addStatut(StatutEnLigne $statut): static
    {
        if (!$this->statut->contains($statut)) {
            $this->statut->add($statut);
            $statut->setUtilisateurId($this);
        }

        return $this;
    }

    public function removeStatut(StatutEnLigne $statut): static
    {
        if ($this->statut->removeElement($statut)) {
            // set the owning side to null (unless already changed)
            if ($statut->getUtilisateurId() === $this) {
                $statut->setUtilisateurId(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection<int, HistoriqueConnexion>
     */
    public function getConnid(): Collection
    {
        return $this->connid;
    }

    public function addConnid(HistoriqueConnexion $connid): static
    {
        if (!$this->connid->contains($connid)) {
            $this->connid->add($connid);
            $connid->setUtilisateurId($this);
        }

        return $this;
    }

    public function removeConnid(HistoriqueConnexion $connid): static
    {
        if ($this->connid->removeElement($connid)) {
            // set the owning side to null (unless already changed)
            if ($connid->getUtilisateurId() === $this) {
                $connid->setUtilisateurId(null);
            }
        }

        return $this;
    }
}
